src/G2/Equiv/Induction.hs

```diff
@@ -0,0 +1,336 @@
+import qualified G2.Language.Typing as T
+import qualified G2.Language.CallGraph as G
+
+import Data.List
+import Data.Maybe
+import qualified Data.Text as DT
+
+import qualified Data.HashSet as HS
+import qualified G2.Solver as S
+
+import qualified G2.Language.PathConds as P
+
+import G2.Equiv.InitRewrite
+import G2.Equiv.EquivADT
+import G2.Equiv.G2Calls
+import G2.Equiv.Tactics
+
+import qualified Data.HashMap.Lazy as HM
+import qualified Data.Map as M
+import G2.Execution.Memory
+import Data.Monoid (Any (..))
+
+import Debug.Trace
+
+import G2.Execution.NormalForms
+import qualified G2.Language.Stack as Stck
+import Control.Monad
+
+import Data.Time
+
+import G2.Execution.Reducer
+import G2.Lib.Printers
+
+import qualified Control.Monad.Writer.Lazy as W
+
+swap :: (a, b) -> (b, a)
+swap (x, y) = (y, x)
+
+otherSide :: Side -> Side
+otherSide ILeft = IRight
+otherSide IRight = ILeft
+
+-- TODO still getting bad induction summary results
+reverseIndMarker :: IndMarker -> IndMarker
+reverseIndMarker im = IndMarker {
+    ind_real_present = swap $ ind_real_present im
+  , ind_used_present = swap $ ind_used_present im
+  , ind_past = swap $ ind_past im
+  , ind_result = swap $ ind_result im
+  , ind_present_scrutinees = swap $ ind_present_scrutinees im
+  , ind_past_scrutinees = swap $ ind_past_scrutinees im
+  , ind_side = otherSide $ ind_side im
+  , ind_fresh_name = ind_fresh_name im
+}
+
+empty_name :: Name
+empty_name = Name (DT.pack "") Nothing 1 Nothing
+
+-- TODO keep track of case statements with no stamp
+readStamps :: Expr -> [Name]
+readStamps (Tick _ e) = readStamps e
+readStamps (Case e i a) =
+  case a of
+    (Alt am1 a1):_ -> case a1 of
+      Tick (NamedLoc n) _ -> n:(readStamps e)
+      _ -> empty_name:(readStamps e)
+    _ -> error "Empty Alt List"
+readStamps _ = []
+
+innerScrutinees :: Expr -> [Expr]
+innerScrutinees (Tick _ e) = innerScrutinees e
+innerScrutinees e@(Case e' _ _) = e:(innerScrutinees e')
+innerScrutinees e = [e]
+
+replaceScrutinee :: Expr -> Expr -> Expr -> Expr
+replaceScrutinee e1 e2 e | e1 == e = e2
+replaceScrutinee e1 e2 (Tick nl e) = Tick nl (replaceScrutinee e1 e2 e)
+replaceScrutinee e1 e2 (Case e i a) = Case (replaceScrutinee e1 e2 e) i a
+replaceScrutinee _ _ e = e
+
+-- checking for depth of first expression within second
+scrutineeDepth :: Expr -> Expr -> Int
+scrutineeDepth e1 e2 | e1 == e2 = 0
+scrutineeDepth e1 (Tick _ e) = scrutineeDepth e1 e
+scrutineeDepth e1 (Case e i a) = 1 + scrutineeDepth e1 e
+scrutineeDepth _ _ = error "Not Contained"
+
+-- the depths do not need to be the same
+-- however, the stamps should match up to the depth of the old one
+validScrutinee :: StateET -> StateET -> StateET -> Bool
+validScrutinee s p pc =
+  let d = scrutineeDepth (exprExtract p) (exprExtract pc)
+      stamps_old = take d $ readStamps (exprExtract pc)
+      stamps_new = take d $ readStamps (exprExtract s)
+  in trace (show (d, stamps_old, stamps_new)) stamps_old == stamps_new
```

Comment:
Can you expand a bit on the comment to explain what it means for a scrutinee to be "valid"?  Also, sort of minor, but pc is not the best name for a State, because we use that all over G2 for path constraints.

Reply:
This function checks the condition for a valid use of induction:  it looks at the stamps for the Case statements in the past and present statements to ensure that they haven't changed.  The first argument s is the present state, p is the state from the past with the scrutinee to be used, and pc ("past container," I think) is the real state from the past that encompasses the scrutinee from the past that's being used.